File: services/tracking/src/features/tracking/api/schemas.py
"""Pydantic v2 response schemas for the REST surface (JE-92/93/94).

## Why these are hand-mapped and not `from_attributes`

The ORM rows carry audit and soft-delete columns (`created_by`, `deleted_at`, …)
that are internal and must never leave the service, so the mapping is explicit —
field by field. The response contract and the table are two different things that
only happen to overlap.

## Timestamps

`datetime` is emitted as an **ISO-8601 string with an explicit `Z`** by `iso()`
below. The columns are naive (MySQL `DATETIME` carries no zone) but hold UTC;
`isoformat()` alone would emit no offset and leave the client guessing.

Serialized as `str`, not as a Pydantic `datetime`: Pydantic would render the naive
value without an offset and silently drop the `Z`.

## Shipping address

Deliberately **absent** from every schema here. It is PII, it is not needed to
render a shipment's status, and the narrowest surface that answers the question is
the one that cannot leak it. It is stored (Orders' snapshot, forwarded at creation)
but never read back out on any surface this service serves.
"""

from __future__ import annotations

from collections.abc import Iterable
from datetime import datetime as _datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from src.features.tracking.domain.models import ID_LENGTH, Tracking, TrackingHistory


def iso(moment: _datetime | None) -> str:
    """Render a DATETIME column as an ISO-8601 string with an explicit `Z`.

    The columns are naive (MySQL DATETIME carries no zone) but hold UTC — the
    repository writes `datetime.now(UTC)` stripped of its tzinfo. `isoformat()` on
    a naive value emits no offset, which would leave the consumer to guess the
    zone, so the `Z` is appended explicitly.

    `None` renders as `""` rather than `null`: every timestamp on this surface is
    NOT NULL in practice, and a string-typed field that can also be null would make
    every client branch on a case that never occurs.
    """
    if moment is None:
        return ""
    return f"{moment.isoformat()}Z"


class HealthResponse(BaseModel):
    """`GET /v1/health` — the exact body the design specifies."""

    status: str = "ok"


class TrackingHistoryEntryResponse(BaseModel):
    """One immutable status transition.

    Carries no `shipping_address` — the address is fixed for a tracking's
    lifetime, so only the tracking itself would hold it (and this surface does not
    expose it at all).
    """

    tracking_id: str
    user_id: str
    order_id: str
    status: str
    datetime: str = Field(description="ISO-8601 UTC timestamp of this transition")

    @classmethod
    def from_entity(cls, entry: TrackingHistory) -> TrackingHistoryEntryResponse:
        return cls(
            tracking_id=entry.tracking_id,
            user_id=entry.user_id,
            order_id=entry.order_id,
            status=entry.status,
            datetime=iso(entry.datetime_),
        )


class TrackingResponse(BaseModel):
    """A tracking together with its ordered history.

    History is part of the payload rather than a separate endpoint because every
    caller of these reads wants both — the design specifies the tracking "+ its
    `Tracking_History`" for the single read and for the batch read alike.
    """

    id: str
    user_id: str
    order_id: str
    status: str
    datetime: str = Field(description="ISO-8601 UTC timestamp of the current status")
    history: list[TrackingHistoryEntryResponse]

    @classmethod
    def from_entity(
        cls, tracking: Tracking, history: Iterable[TrackingHistory]
    ) -> TrackingResponse:
        """Map a row plus its history.

        `history` is passed in rather than read off `tracking.history` for the same
        reason `tracking_with_history_to_proto` does it: the caller stays in charge
        of where the ordering came from. Both callers hand over the relationship,
        which sorts by `TrackingHistory.ordering()` — timestamp, then progression
        position. A bare timestamp sort ties on same-second transitions and MySQL
        then falls back to PK order, which is alphabetical and puts DELIVERED first.
        """
        return cls(
            id=tracking.id,
            user_id=tracking.user_id,
            order_id=tracking.order_id,
            status=tracking.status,
            datetime=iso(tracking.datetime_),
            history=[
                TrackingHistoryEntryResponse.from_entity(entry) for entry in history
            ],
        )


class TrackingListResponse(BaseModel):
    """The batch read's payload.

    An object with a `trackings` list rather than a bare top-level array: a bare
    array is not extensible (there is nowhere to add a field later without a
    breaking change) and is the shape most REST clients handle worst.

    There is no `total`, and no per-id error entry. Non-owned and unknown ids are
    silently omitted per the design's ownership rule, so a count of what came back
    is exactly `len(trackings)` and anything more would start describing what the
    caller does NOT own.
    """

    trackings: list[TrackingResponse]


class InitTrackingRequest(BaseModel):
    """Body of `POST /v1/trackings/init-tracking` (JE-105).

    ## Two fields, and deliberately no identity

    `order_id` and `shipping_address`, nothing else. The caller's identity is NOT
    in the body and must never be: it arrives as the gateway-injected `x-user-id`
    header, which the gateway derives from a verified Cognito JWT. A `user_id` (or
    `cognito_sub`) field here would be an unauthenticated string a client chooses,
    so anyone could create a tracking attributed to anyone — the body is client
    input, the header is a gateway assertion. The service resolves the internal
    `usr_` id itself, through Users, rather than believing a claim it was handed.

    ## `test_mode` is not here either

    It travels as the `x-test-mode` header — see the router's docstring for why.

    `model_config` forbids extra fields: a client sending `user_id` gets a `422`
    naming the field, rather than having it silently ignored and later wondering
    why the tracking belongs to someone else.
    """

    model_config = ConfigDict(extra="forbid")

    order_id: str = Field(
        min_length=1,
        max_length=ID_LENGTH,
        description="The order this tracking follows.",
    )

    #: The delivery address snapshot, free-form JSON. PII — never logged.
    #:
    #: Typed as a permissive mapping rather than a strict `Address` model on
    #: purpose: this column is a point-in-time SNAPSHOT whose shape is owned by the
    #: caller that resolved it (Orders, from Users' `Address` message), and the
    #: service neither reads it back nor renders it on any surface. A strict model
    #: here would make Tracking reject an address the moment Users adds a field,
    #: turning an additive upstream change into a creation outage — for data this
    #: service only stores. Empty/None is accepted: an order whose address the
    #: caller could not resolve must still be trackable.
    shipping_address: dict[str, Any] | None = Field(
        default=None,
        description="Point-in-time delivery address snapshot. Never logged (PII).",
    )


class InitTrackingResponse(BaseModel):
    """`201` payload: the created tracking, at `PLACED`, with its first history row.

    Reuses `TrackingResponse`'s shape rather than declaring a leaner one, so the
    body a client gets from creating a tracking is identical to the one it gets
    from reading it back — including the deliberate omissions (`shipping_address`
    and `cognito_sub` are PII/identity and appear on neither).
    """

    tracking: TrackingResponse


class UpdateStatusRequest(BaseModel):
    """Body of `PUT /v1/trackings/{order_id}/status`.

    `status` is validated as a *string* here and parsed into `TrackingStatus` in
    the handler via `parse_status`. Declaring it as the enum would let Pydantic
    reject an unknown value with a 422 before the handler ran — but the design
    specifies `400` for a rejected status, and routing invalid-value handling
    through one place keeps the four failure reasons (unknown value plus the three
    transition guards) answering with the same status code and shape.
    """

    status: str


class E2eCleanupResponse(BaseModel):
    """`200` payload of `DELETE /v1/trackings/e2e-cleanup`.

    One field, and it earns its place: "the suite still sees its fixtures" and
    "the cleanup matched nothing" look identical from the harness's side, so a
    bodiless `204` would leave it unable to tell a teardown that worked from one
    that silently selected zero rows without reading this service's logs. Users'
    cleanup returns its `count` for the same reason.

    Named `deleted` rather than `count` because it says what was counted.
    """

    deleted: int


class ErrorResponse(BaseModel):
    """Failure payload.

    `reason` is the machine-readable `TransitionRejectionReason`
    (`already_delivered` / `backward_transition` / `not_strictly_forward`) or a
    comparable code — the same value the `*_failed` log line carries, so a caller
    debugging a rejection and an operator reading the logs see the same token.
    """

    detail: str
    reason: str

File: services/tracking-go/internal/openapi/schemas.go
package openapi

// The component schemas.
//
// # Reproduced from the Python contract, INCLUDING its Pydantic artifacts
//
// Every `title` here is one Pydantic auto-derived from a field name, and no
// consumer branches on a title. They are kept anyway: dropping them would be a
// difference on every schema at once, and the allowlist would then be carrying an
// entry that hides thirteen real ones behind a wildcard. Copying them costs
// nothing and keeps the diff genuinely empty.
//
// # Two schemas here are NOT in the Python document, deliberately
//
// NestedErrorResponse and NestedErrorBody describe the body init-tracking's 404
// and 409 actually serve. The Python CODE raises
// HTTPException(detail={"detail": ..., "reason": ...}), which FastAPI renders as
// {"detail": {"detail": ..., "reason": ...}} -- nested. FastAPI's generator
// cannot express that wrapping and emits the flat ErrorResponse instead, so THE
// PYTHON SPEC IS WRONG THERE AND THE PYTHON CODE IS RIGHT. The Go handler matches
// the code, this document matches the Go handler, and allowlist.go records the
// difference from the spec rather than "fixing" the handler to match a body no
// deployed service has ever served.
//
// # What is absent, and must stay absent
//
// No response schema carries `shipping_address` (PII) or `cognito_sub`
// (identity). `datetime` is a plain `string` on both response schemas -- the wire
// value is isoformat() + "Z" with microsecond precision, and "" when absent. It
// is NOT `format: date-time`, which would tell a generated client to expect
// RFC3339 and to parse "" as an error, and it is NOT nullable, because the field
// is never null.

// schemas returns components.schemas.
func schemas() map[string]any {
	return map[string]any{
		"E2eCleanupResponse": map[string]any{
			"properties": map[string]any{
				"deleted": map[string]any{
					"type":  "integer",
					"title": "Deleted",
				},
			},
			"type": "object",
			"required": []any{
				"deleted",
			},
			"title": "E2eCleanupResponse",
			"description": "`200` payload of `DELETE /v1/trackings/e2e-cleanup`.\n" +
				"\n" +
				"One field, and it earns its place: \"the suite still sees its fixtures\" and\n" +
				"\"the cleanup matched nothing\" look identical from the harness's side, so a\n" +
				"bodiless `204` would leave it unable to tell a teardown that worked from one\n" +
				"that silently selected zero rows without reading this service's logs. Users'\n" +
				"cleanup returns its `count` for the same reason.\n" +
				"\n" +
				"Named `deleted` rather than `count` because it says what was counted.",
		},
		"ErrorResponse": map[string]any{
			"properties": map[string]any{
				"detail": map[string]any{
					"type":  "string",
					"title": "Detail",
				},
				"reason": map[string]any{
					"type":  "string",
					"title": "Reason",
				},
			},
			"type": "object",
			"required": []any{
				"detail",
				"reason",
			},
			"title": "ErrorResponse",
			"description": "Failure payload.\n" +
				"\n" +
				"`reason` is the machine-readable `TransitionRejectionReason`\n" +
				"(`already_delivered` / `backward_transition` / `not_strictly_forward`) or a\n" +
				"comparable code — the same value the `*_failed` log line carries, so a caller\n" +
				"debugging a rejection and an operator reading the logs see the same token.",
		},
		"HTTPValidationError": map[string]any{
			"properties": map[string]any{
				"detail": map[string]any{
					"items": map[string]any{
						"$ref": "#/components/schemas/ValidationError",
					},
					"type":  "array",
					"title": "Detail",
				},
			},
			"type":  "object",
			"title": "HTTPValidationError",
		},
		"HealthResponse": map[string]any{
			"properties": map[string]any{
				"status": map[string]any{
					"type":    "string",
					"title":   "Status",
					"default": "ok",
				},
			},
			"type":        "object",
			"title":       "HealthResponse",
			"description": "`GET /v1/health` — the exact body the design specifies.",
		},
		"InitTrackingRequest": map[string]any{
			"properties": map[string]any{
				"order_id": map[string]any{
					"type":        "string",
					"maxLength":   28,
					"minLength":   1,
					"title":       "Order Id",
					"description": "The order this tracking follows.",
				},
				"shipping_address": map[string]any{
					"anyOf": []any{
						map[string]any{
							"type": "object",
						},
						map[string]any{
							"type": "null",
						},
					},
					"title":       "Shipping Address",
					"description": "Point-in-time delivery address snapshot. Never logged (PII).",
				},
			},
			"additionalProperties": false,
			"type":                 "object",
			"required": []any{
				"order_id",
			},
			"title": "InitTrackingRequest",
			"description": "Body of `POST /v1/trackings/init-tracking` (JE-105).\n" +
				"\n" +
				"## Two fields, and deliberately no identity\n" +
				"\n" +
				"`order_id` and `shipping_address`, nothing else. The caller's identity is NOT\n" +
				"in the body and must never be: it arrives as the gateway-injected `x-user-id`\n" +
				"header, which the gateway derives from a verified Cognito JWT. A `user_id` (or\n" +
				"`cognito_sub`) field here would be an unauthenticated string a client chooses,\n" +
				"so anyone could create a tracking attributed to anyone — the body is client\n" +
				"input, the header is a gateway assertion. The service resolves the internal\n" +
				"`usr_` id itself, through Users, rather than believing a claim it was handed.\n" +
				"\n" +
				"## `test_mode` is not here either\n" +
				"\n" +
				"It travels as the `x-test-mode` header — see the router's docstring for why.\n" +
				"\n" +
				"`model_config` forbids extra fields: a client sending `user_id` gets a `422`\n" +
				"naming the field, rather than having it silently ignored and later wondering\n" +
				"why the tracking belongs to someone else.",
		},
		"InitTrackingResponse": map[string]any{
			"properties": map[string]any{
				"tracking": map[string]any{
					"$ref": "#/components/schemas/TrackingResponse",
				},
			},
			"type": "object",
			"required": []any{
				"tracking",
			},
			"title": "InitTrackingResponse",
			"description": "`201` payload: the created tracking, at `PLACED`, with its first history row.\n" +
				"\n" +
				"Reuses `TrackingResponse`'s shape rather than declaring a leaner one, so the\n" +
				"body a client gets from creating a tracking is identical to the one it gets\n" +
				"from reading it back — including the deliberate omissions (`shipping_address`\n" +
				"and `cognito_sub` are PII/identity and appear on neither).",
		},
		"InternalDeleteByUserRequest": map[string]any{
			"properties": map[string]any{
				"cognito_sub": map[string]any{
					"type":      "string",
					"minLength": 1,
					"title":     "Cognito Sub",
				},
				"user_id": map[string]any{
					"type":      "string",
					"minLength": 1,
					"title":     "User Id",
				},
			},
			"type": "object",
			"required": []any{
				"cognito_sub",
				"user_id",
			},
			"title":       "InternalDeleteByUserRequest",
			"description": "Body of `DELETE /v1/trackings/by-user`. Both identities are required.",
		},
		"InternalDeleteByUserResponse": map[string]any{
			"properties": map[string]any{
				"deleted": map[string]any{
					"type":  "integer",
					"title": "Deleted",
				},
			},
			"type": "object",
			"required": []any{
				"deleted",
			},
			"title": "InternalDeleteByUserResponse",
			"description": "`200` payload of `DELETE /v1/trackings/by-user`.\n" +
				"\n" +
				"Named `deleted` rather than `count` because it says what was counted — the same\n" +
				"shape Users and Orders report.",
		},
		"TrackingHistoryEntryResponse": map[string]any{
			"properties": map[string]any{
				"tracking_id": map[string]any{
					"type":  "string",
					"title": "Tracking Id",
				},
				"user_id": map[string]any{
					"type":  "string",
					"title": "User Id",
				},
				"order_id": map[string]any{
					"type":  "string",
					"title": "Order Id",
				},
				"status": map[string]any{
					"type":  "string",
					"title": "Status",
				},
				"datetime": map[string]any{
					"type":        "string",
					"title":       "Datetime",
					"description": "ISO-8601 UTC timestamp of this transition",
				},
			},
			"type": "object",
			"required": []any{
				"tracking_id",
				"user_id",
				"order_id",
				"status",
				"datetime",
			},
			"title": "TrackingHistoryEntryResponse",
			"description": "One immutable status transition.\n" +
				"\n" +
				"Carries no `shipping_address` — the address is fixed for a tracking's\n" +
				"lifetime, so only the tracking itself would hold it (and this surface does not\n" +
				"expose it at all).",
		},
		"TrackingListResponse": map[string]any{
			"properties": map[string]any{
				"trackings": map[string]any{
					"items": map[string]any{
						"$ref": "#/components/schemas/TrackingResponse",
					},
					"type":  "array",
					"title": "Trackings",
				},
			},
			"type": "object",
			"required": []any{
				"trackings",
			},
			"title": "TrackingListResponse",
			"description": "The batch read's payload.\n" +
				"\n" +
				"An object with a `trackings` list rather than a bare top-level array: a bare\n" +
				"array is not extensible (there is nowhere to add a field later without a\n" +
				"breaking change) and is the shape most REST clients handle worst.\n" +
				"\n" +
				"There is no `total`, and no per-id error entry. Non-owned and unknown ids are\n" +
				"silently omitted per the design's ownership rule, so a count of what came back\n" +
				"is exactly `len(trackings)` and anything more would start describing what the\n" +
				"caller does NOT own.",
		},
		"TrackingResponse": map[string]any{
			"properties": map[string]any{
				"id": map[string]any{
					"type":  "string",
					"title": "Id",
				},
				"user_id": map[string]any{
					"type":  "string",
					"title": "User Id",
				},
				"order_id": map[string]any{
					"type":  "string",
					"title": "Order Id",
				},
				"status": map[string]any{
					"type":  "string",
					"title": "Status",
				},
				"datetime": map[string]any{
					"type":        "string",
					"title":       "Datetime",
					"description": "ISO-8601 UTC timestamp of the current status",
				},
				"history": map[string]any{
					"items": map[string]any{
						"$ref": "#/components/schemas/TrackingHistoryEntryResponse",
					},
					"type":  "array",
					"title": "History",
				},
			},
			"type": "object",
			"required": []any{
				"id",
				"user_id",
				"order_id",
				"status",
				"datetime",
				"history",
			},
			"title": "TrackingResponse",
			"description": "A tracking together with its ordered history.\n" +
				"\n" +
				"History is part of the payload rather than a separate endpoint because every\n" +
				"caller of these reads wants both — the design specifies the tracking \"+ its\n" +
				"`Tracking_History`\" for the single read and for the batch read alike.",
		},
		"UpdateStatusRequest": map[string]any{
			"properties": map[string]any{
				"status": map[string]any{
					"type":  "string",
					"title": "Status",
				},
			},
			"type": "object",
			"required": []any{
				"status",
			},
			"title": "UpdateStatusRequest",
			"description": "Body of `PUT /v1/trackings/{order_id}/status`.\n" +
				"\n" +
				"`status` is validated as a *string* here and parsed into `TrackingStatus` in\n" +
				"the handler via `parse_status`. Declaring it as the enum would let Pydantic\n" +
				"reject an unknown value with a 422 before the handler ran — but the design\n" +
				"specifies `400` for a rejected status, and routing invalid-value handling\n" +
				"through one place keeps the four failure reasons (unknown value plus the three\n" +
				"transition guards) answering with the same status code and shape.",
		},
		"ValidationError": map[string]any{
			"properties": map[string]any{
				"loc": map[string]any{
					"items": map[string]any{
						"anyOf": []any{
							map[string]any{
								"type": "string",
							},
							map[string]any{
								"type": "integer",
							},
						},
					},
					"type":  "array",
					"title": "Location",
				},
				"msg": map[string]any{
					"type":  "string",
					"title": "Message",
				},
				"type": map[string]any{
					"type":  "string",
					"title": "Error Type",
				},
			},
			"type": "object",
			"required": []any{
				"loc",
				"msg",
				"type",
			},
			"title": "ValidationError",
		},

		// --- Not in the Python document. See the package comment above. ---
		"NestedErrorBody": map[string]any{
			"properties": map[string]any{
				"detail": map[string]any{
					"type":  "string",
					"title": "Detail",
				},
				"reason": map[string]any{
					"type":  "string",
					"title": "Reason",
				},
			},
			"type": "object",
			"required": []any{
				"detail",
				"reason",
			},
			"title": "NestedErrorBody",
			"description": "The inner object of `NestedErrorResponse`.\n" +
				"\n" +
				"Field-for-field identical to `ErrorResponse`, and a SEPARATE schema on\n" +
				"purpose: the two occupy different positions on the wire (`ErrorResponse` is\n" +
				"the whole body of the carrier 400, this one sits under `detail`), so sharing\n" +
				"a name would make the nesting invisible to anyone reading the document.",
		},
		"NestedErrorResponse": map[string]any{
			"properties": map[string]any{
				"detail": map[string]any{
					"$ref": "#/components/schemas/NestedErrorBody",
				},
			},
			"type": "object",
			"required": []any{
				"detail",
			},
			"title": "NestedErrorResponse",
			"description": "The body `POST /v1/trackings/init-tracking` serves on 404 and 409.\n" +
				"\n" +
				"NESTED, and the only place on this surface that shape occurs: the detail is\n" +
				"an OBJECT carrying its own `detail` and `reason`, not a string. The Python\n" +
				"raises `HTTPException(detail={\"detail\": ..., \"reason\": ...})` and FastAPI\n" +
				"wraps a structured detail this way, so this is what every deployed client has\n" +
				"always received -- even though the generated Python spec declares a flat\n" +
				"`ErrorResponse` here, because FastAPI cannot express the wrapping.",
		},
	}
}
